### PR TITLE
fix panic: concurrent write to websocket connection

### LINKED ISSUE #3907: [api] concurrent write to websocket connection
Found that if the data is sent more frequently, it will lead to data write concurrency with `ping` service, it will cause websocket concurrent write panic.

the Bug is because  the package `github.com/gorilla/websocket`  does not support concurrent writers

https://github.com/gorilla/websocket/issues/119

### PR DESCRIPTION
**Pull request by millken**
# Description
as title

Fixes #3907

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/websocket.go [hunk 1]
```diff
@@ -36,6 +37,27 @@ var upgrader = websocket.Upgrader{
...
+	return c.ws.WriteJSON(message)
+}
+
+// WriteMessage writes a message to the connection in a thread-safe way
+func (c *safeWebsocketConn) WriteMessage(messageType int, data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.ws.WriteMessage(messageType, data)
+}
+
```
Comment (dustinxie) [inline api/websocket.go:60]: add `Close()` and `SetDeadline`

File: api/websocket.go [hunk 2]
```diff
@@ -36,6 +37,41 @@ var upgrader = websocket.Upgrader{
 	WriteBufferSize: 1024,
 }
 
+// type safeWebsocketConn wraps websocket.Conn with a mutex
```
Comment (Liuhaai) [inline api/websocket.go:40]: the link could be added here 
https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

**Comment by Liuhaai** [comment]
good finding

File: api/websocket.go [hunk 3]
```diff
@@ -36,6 +37,41 @@ var upgrader = websocket.Upgrader{
...
+
+// WriteMessage writes a message to the connection in a thread-safe way
+func (c *safeWebsocketConn) WriteMessage(messageType int, data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.ws.WriteMessage(messageType, data)
+}
+
+// Close closes the underlying network connection without sending or waiting for a close frame
+func (c *safeWebsocketConn) Close() error {
```
Comment (Liuhaai) [inline api/websocket.go:62]: `Close()` isn't required for atomic op

**Review by dustinxie — APPROVED** [review]

File: api/websocket.go [hunk 4]
```diff
@@ -70,7 +104,8 @@ func (wsSvr *WebsocketHandler) handleConnection(ctx context.Context, ws *websock
 	})
 
 	ctx, cancel := context.WithCancel(ctx)
-	go ping(ctx, ws, cancel)
+	safeWs := &safeWebsocketConn{ws: ws}
+	go ping(ctx, safeWs, cancel)
```
Comment (dustinxie) [inline api/websocket.go:108]: 1. according to https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency, Reader method also need concurrent control, like `SetReadDeadline`,  `SetReadLimit`
2. L95-103 above, `SetReadDeadline` called 2 times, is that a duplicate?

Comment (millken) [reply api/websocket.go:108]: 1. see websocket code, panic is only triggered on write(WriteMessage, WriteJSON), Set* will not
```go
	if c.isWriting {
		panic("concurrent write to websocket connection")
	}
```

2. the first one is used to set the initial read deadline, ensuring that data is read from the WebSocket connection within the specified time. The second one is used to reset the read deadline after receiving a heartbeat message.

Comment (dustinxie) [reply api/websocket.go:109]: but pkg.go.dev clearly stated:

"Applications are responsible for ensuring that no more than one goroutine calls the write methods (NextWriter, SetWriteDeadline, WriteMessage, WriteJSON, EnableWriteCompression, SetCompressionLevel) concurrently and **that no more than one goroutine calls the read methods** (NextReader, SetReadDeadline, ReadMessage, ReadJSON, SetPongHandler, SetPingHandler) concurrently."

**Pull request merged**